### DISCUSSION TITLE
Unstuck RAW iCloud sync

### DISCUSSION THREAD

**Discussion by PhilipTrauner** [RhetTbull/osxphotos#1181] [Q&A]
I've consistently managed to break iCloud photo sync when importing (built-in Photos app import functionality) RAW photos.

The only reliable workaround to unstuck iCloud sync I've discovered so far, is to [rotate all photos that are _likely_ stuck, followed by undoing the rotation again](https://www.reddit.com/r/MacOS/comments/yyu1ku/photos_app_stuck_not_syncing_photos_to_icloud/).

`osxphoto inspect` reveals that the workaround results in the "Date modified" field being set for photos it was applied to, which presumably fixes something in the syncing machinery.

`osxphotos` has already helped me tremendously by listing which photos are _actually_ stuck (`osxphotos query --not-incloud --not-shared`), as rotating unaffected photos, which have been edited externally, sometimes corrupts their sidecar file. 

Now I was wondering if there's a way that `osxphotos` could be used **instead** of the workaround. 

I already tried using your (pretty awesome) [script to set location data based on a `.gpx` file](https://github.com/RhetTbull/add_photo_locations_from_gpx), in hopes that a metadata change would unstuck syncing, but that doesn't appear to be the case.

Is there any operation which could potentially trigger a sync?
More than willing to share my photo database in case that's useful 😇

**Comment by RhetTbull** [comment]
I don't know how to force iCloud to upload the photo but if you change the time, it can force a metadata sync. (Though this might not work in your case as it sounds like you already tried changing metadata). If you want to give it a try, select an offending photo then run:

`osxphotos timewarp --time-delta "+1s"` 

This adds a second to the time. You can then immediately run the reverse:

`osxphotos timewarp --time-delta "-1s"`

Which restores the time to what it was. 

I'm glad the gpx script was helpful! I plan to incorporate that functionality into the `osxphotos add-locations` command.

**Reply by RhetTbull** [reply]
If that doesn't work, I'm working on a feature to allow osxphotos to perform external edits on the library. It might be able to force the same thing as the rotate does using a workflow like this:

If photo has an external edit, save it and the sidecar file. 
Perform an edit that effectively does nothing other than tell Photos the photo was edited. 
Restore the original edit. 

This is a ways off as it requires hacking some undocumented features of Photos but I'm slowly making progress.

**Reply by PhilipTrauner** [reply]
Just tried `timewarp`, and your hunch was sadly correct, syncing does not appear to be happening 😔
I think that its unreasonable to expect folks to come up with workarounds for something that does legitimately appear to be a relatively easy to hit bug in Photos (and / or `bird`).

I've scheduled a call with Apple Support, and I'll try to submit a Radar, but I'm afraid that I don't have sufficient context yet, because the iCloud syncing machinery is so opaque. 
Even `osxphotos query --not-incloud --not-shared` isn't listing stuck photos anymore 😣
`sudo brctl log -w` also isn't printing out anything in response to external edits and rotation.

Really sorry that I'm bothering you with this, but do you know of any other incantations beside these that could yield more context?

**Reply by RhetTbull** [reply]
Unfortunately I can't think of anything else to force syncing. It is frustrating there's no way to easily figure out what's going on with iCloud. I'm willing to take a look if you want to send some debug data.  It's *possible* there's something we could tweak in the database that would reset sync status (I've not delved into this) but I think it's a long shot.

In the All Photos or Library view, select one of the photos that is stuck and run the following command:

`osxphotos debug-dump --dump photos --selected > bad.txt`

Then select another photo you've imported the same way that has actually synced and run the command again:

`osxphotos debug-dump --dump photos --selected > good.txt`

That will dump a bunch of data from the database that might offer some clues.  You can post the data here (it may contain locations, etc. that you might consider sensitive) or send to me at rturnbull+git@gmail.com

**Reply by PhilipTrauner** [reply]
Here you go, thank you so much for doing this 😇

[bad.txt](https://github.com/RhetTbull/osxphotos/files/12448841/bad.txt)
[good.txt](https://github.com/RhetTbull/osxphotos/files/12448842/good.txt)

**Reply by RhetTbull** [reply]
One clue:

Good:
```
ZASSET:
        'ZCLOUDLOCALSTATE': 1,
        'ZCLOUDPLACEHOLDERKIND': 4,
        'ZDEFERREDPROCESSINGNEEDED': 0,
```

Bad:
```
ZASSET:
        'ZCLOUDLOCALSTATE': 0,
        'ZCLOUDPLACEHOLDERKIND': 0,
        'ZDEFERREDPROCESSINGNEEDED': 4,
```

For the "bad" photo, `ZCLOUDLOCALSTATE =0` indicates the photo is not synced to iCloud (which you know). What's interesting is the value `ZDEFERREDPROCESSINGNEEDED=4` which is `0` for the "good" photo. I've not explored `ZDEFERREDPROCESSINGNEEDED` but in my library of 36k photos, there isn't a single one where this value is not 0. I'm wondering if that column means Photos thinks it needs to do something to the photo before uploading to iCloud?

Try the following commands.  At the terminal, run:

```bash
sqlite3 ~/Pictures/Photos\ Library.photoslibrary/database/Photos.sqlite
```

Then type the following:

```sql
SELECT COUNT(*) FROM ZASSET WHERE ZDEFERREDPROCESSINGNEEDED <> 0;
```

Finally, you can press `Ctrl+D` to exit the sqlite prompt.

Let me know the count. I get `0`:

```
sqlite> SELECT COUNT(*) FROM ZASSET WHERE ZDEFERREDPROCESSINGNEEDED <> 0;
0
```

If your count is close to the number of "stuck" photos, we may be on to something. Will just need to figure out what this means or maybe try to change one and see if it gets "unstuck".

**Reply by PhilipTrauner** [reply]
I currently have a single stuck photo in my library, and the query does indeed return 1 🙌

**Reply by RhetTbull** [reply]
Definitely a good clue then. Unfortunately I'm not sure what to do about it. How long as the photo been "stuck"? 

A few more data points:

There's a "cloud metadata" record stored in `ZCLOUDMASTERMEDIAMETADATA` table. The "good" photo has this record completed as expected (below is a snippet of what's in there) but the "bad" photo has no record in this table. It may be this is computed on the cloud but I believe that nearly all processing of photos is done "on device" as Apple is very privacy sensitive.  It could be some process that's supposed to update this metadata hasn't completed or crashed (and could be linked to the `ZDEFERREDPROCESSINGNEEDED` value). 

```
'{DNG}': {
            'DNGBackwardVersion': [1, 2],
            'PreviewColorSpace': 2,
            'AnalogBalance': [1.0, 1.0, 1.0],
            'CalibrationIlluminant1': 17,
            'BaselineExposure': 0.3064727783203125,
            'BaselineSharpness': 1.0,
            'CalibrationIlluminant2': 21,
            'WhiteLevel': [14328],
            'BaselineNoise': 1.0,
            'BlackLevelRepeatDim': [2, 2],
            'ColorMatrix2': [
```

Other clues: 

`ZASSET.ZMASTER` is null. This should be the foreign key to the `ZCLOUDMASTER` record for the photo. But it might be expected that's it null as there is no master photo in the cloud as the photo's not in the cloud.

`ZASSET.ZMEDIAANALYSISATTRIBUTES` is null. This is a foreign key to another table so whatever fills this in hasn't completed. (Maybe `photoanalysisd` though this isn't one of the values osxphotos currently uses to get the analysis attributes (category labels, etc.)). `ZASSET.ZPHOTOANALYSISATTRIBUTES` is not null so at least some analysis has been completed on the photo (this is also a foreign key). 

I'll keep searching. It might be possible to manually change the value of `ZDEFERREDPROCESSINGNEEDED` to see if that unsticks it but there's a risk of corrupting something in the library as we don't fully understand what that means. 

If this photo (`R0000443.DNG`) is one that you can share, would you mind sending to me at rturnbull+git@gmail.com and I will import it into a test library and monitor the database activity to see if it is perhaps something unique to these RAW photos themselves or something else.

**Reply by PhilipTrauner** [reply]
Just sent you a mail, some more findings from my end: Editing inside the Photos app and consequently undoing all edits also seems to cause the syncing machinery to kick into gear, which would suggest that adding and consequently removing a sidecar file also does _something_ 😄
